rfd/0201-native-auto-approval.md

```diff
@@ -0,0 +1,364 @@
+use cases, Teleport should support automatic approvals natively.
+
+Automatic approvals enables teams to enforce zero standing privileges, while
+allowing users to get access to their pre approved resources for a limited
+period of time. There are alternative ways to enforce zero standing privileges,
+but a big part of this feature request is to provide a frictionless user
+experience. For example, access lists can be used to achieve similar behavior,
+but a limitation with access lists is that the user must logout and login again
+to renew their certificates to obtain the new roles.
+
+This feature has also been requested for internal use at Teleport. The Tools
+team enforces gated access to certain pipeline activity. The team would like the
+dev environment to closely mimic the production environment and would like to
+enforce the same gated access, but it would reduce a lot of friction if access
+requests can be automatically approved in the dev environment.
+
+## Goals
+1. Native support for automatic approvals. Integration with external service is
+not required.
+2. Automatic approvals can be configured to allow users with certain traits to
+be automatically approved. For example, all `L1` engineers on team `Cloud` in
+location `Seattle` are pre-approved for the access request.
+3. Automatic approvals can be configured to approve access to select resources.
+For example, the requesting user is pre-approved to access all resources in the
+`dev` environment.
+4. Automatic approvals can be configured using Access Monitoring Rules.
+5. Automatic approvals can be configured using the Teleport Web UI.
+6. Automatic approval rules are easily reviewable through the Teleport Web UI.
+7. User experience is the focus. The feature should be easy to use and configure.
+8. Implementation should be compatible with future plugin interface refactoring.
+
+Note: Access monitoring rules do not currently support access requests by
+resources, in any capacity. It will require more than just extending the access
+monitoring rule predicate language to support this use case. Goal [3] will be
+left out of scope for this RFD. It probably deserves it's own separate RFD. As a
+workaround to support this use case, auto approval rules can be configured to
+approve a role that only has the permission to access resources in a specific
+environment.
+
+Note: Development is in progress to refactor the access plugins and implement a
+unified set of interfaces. This will help achieve feature parity across access
+plugins. Auto approval support for other plugins is out of scope for this RFD,
+but it is something to consider while we decide how to implement native auto
+approvals. This is why [8] is included in the goals list. See
+https://github.com/gravitational/teleport/issues/47150 for more details.
+
+## Details
+This feature will be supported by a new `access-native` plugin. This native
+plugin will implement the same interface as already existing access plugins.
```

Comment:
It's a bit odd for the RFD to start with "users don't want to manage auto approval via a plugin" and have the proposed solution be yet another plugin. 

If it's another plugin, why are we calling it "native" auto approval? To me, native suggests built-in and available by default.

Reply:
It does sound contradictory 😅.

The point I was trying to convey with "users don't want to manage auto approval via a plugin" is that we allow users to build their own access request plugins with Teleport. So they could technically build and support this plugin themselves. I think this is what we had suggested to a customer, but they did not want that responsibility.

Whether this feature is supported via the plugin interface or built-in to Teleport directly is more of an implementation detail. It seemed easier to just implement the already existing plugin interface, and then have the plugin always run by default.

Reply:
> It seemed easier to just implement the already existing plugin interface, and then have the plugin always run by default.

I'm not sure I agree. We have lots of functionality in Teleport that always runs by default (auto discovery, notification generation, backend expiry, upload completer, etc) but we don't implement the plugin interface for these things. (Because they're not plugins, they're just features of Teleport..)

Reply:
I don't have a very strong opinion on whether this feature should be implemented using the plugin interface. But I do think implementing the plugin interface will be supporting our goal for feature parity https://github.com/gravitational/teleport/issues/47150. Extending or modifying auto approvals will be easier if this feature is also implemented using the plugin interface.

What would be some benefits to consider by not implementing the plugin interface?